server/src/main/java/org/elasticsearch/index/snapshots/blobstore/SnapshotFiles.java

```diff
@@ -48,9 +52,25 @@ public String snapshot() {
      * @param snapshot   snapshot name
      * @param indexFiles index files
      */
-    public SnapshotFiles(String snapshot, List<FileInfo> indexFiles ) {
+    public SnapshotFiles(String snapshot, List<FileInfo> indexFiles, long globalCheckpoint, String historyUUID) {
         this.snapshot = snapshot;
         this.indexFiles = indexFiles;
+        this.globalCheckpoint = globalCheckpoint;
+        this.historyUUID = historyUUID;
+    }
+
+    /**
+     * Returns the shard's global checkpoint at the time the snapshot was taken
+     */
+    public long globalCheckpoint() {
```

Comment:
I'm not sure whether we should call this the global checkpoint. We're only supposed to record this if global checkpoint == max sequence number. I would rather introduce a new term here.
(same for other occurrences of "globalCheckpoint")

Reply:
How about "safeCheckpoint" with appropriate Javadoc in `Repository` and then compute that number in `SnapshotsShardsService` and padd down + store a `-2` if the three sequence numbers don't match up?

Reply:
I simplified this a lot now. The checkpoint isn't explicitly used in the repository code anymore as will be explained below.